extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java

```diff
@@ -87,12 +81,27 @@ public Object get(ByteBuffer buf, int position)
   //Note that this is not threadsafe and I don't think it needs to be
   private Union getUnion(ByteBuffer buf, int position)
   {
-    Union union = unions.get(position);
-    if (union == null) {
-      Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
-      union = (Union) SetOperation.wrap(mem);
-      unions.put(position, union);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
+    Union union = unionMap != null ? unionMap.get(position) : null;
+    if (union != null) {
+      return union;
+    }
+    return createNewUnion(buf, position, true);
+  }
+
+  private Union createNewUnion(ByteBuffer buf, int position, boolean isWrapped)
+  {
+    NativeMemory nm = getNativeMemory(buf);
+    Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
+    Union union = isWrapped
+                  ? (Union) SetOperation.wrap(mem)
+                  : (Union) SetOperation.builder().initMemory(mem).build(size, Family.UNION);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
```

Comment:
@gianm 

>  I guess your idea is that the contract would be something like this: ...

Yes, exactly.

> I'm still willing to defer to @akashdw whether or not this optimization is necessary right now. If the SketchAggregator.updateUnion is expected to take seriously longer than getUnion then the optimization might not yield a noticeable performance difference.

I don't consider this as optimization that may or may not be implemented by @akashdw. I agree with @gianm that performance impact is likely small. However instead of addressing my original small code quality comment (from last week) that would have taken less than 5 min, @akashdw started a dispute and then a super time-draining benchmark battle, pretending that performance difference is not small, inexplicably (for me) reluctant to making that small fix. Given that this PR blocks 0.10.0 release, that supposedly @akashdw needs it to go out, because he reported the bug, but not me (Metamarkets), because we anyway couldn't upgrade to 0.10.0, we are waiting for 0.10.1.

I'm not going to take responsibility for this.